src/go/guestagent/pkg/iptables/iptables_test.go

```diff
@@ -0,0 +1,267 @@
+					id := <-testTracker.receivedRemoveID
+					expectedID := utils.GenerateID(entryToString(removedEntries[i]))
+					require.Equal(t, expectedID, id)
+				}
+
+				addedElement := updatedEntries[len(updatedEntries)-1]
+				id := <-testTracker.receivedID
+				expectedID := utils.GenerateID(entryToString(addedElement))
+				require.Equal(t, expectedID, id)
+
+				pm := <-testTracker.receivedPortMapping
+				portProto, err := nat.NewPort("tcp", strconv.Itoa(addedElement.Port))
+				require.NoError(t, err)
+
+				expectedPortMap := nat.PortMap{
+					portProto: []nat.PortBinding{
+						{
+							HostIP:   tt.listenerIP.String(),
+							HostPort: strconv.Itoa(addedElement.Port),
+						},
+					},
+				}
+				require.ElementsMatch(t, pm[portProto], expectedPortMap[portProto])
+			}
+		})
+	}
+}
+
+func TestForwardPortsSamePortDifferentIP(t *testing.T) {
+	duplicatedPort := 1084
+	tests := []struct {
+		name               string
+		listenerIP         net.IP
+		expectedEntries    []limaIptables.Entry
+		expectedAddFuncErr error
+	}{
+		{
+			name:       "Same Port with different IP",
+			listenerIP: net.IPv4(0, 0, 0, 0),
+			expectedEntries: []limaIptables.Entry{
+				{TCP: true, IP: net.IPv4(192, 168, 22, 10), Port: 1080},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 11), Port: 1081},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 12), Port: 1082},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 13), Port: 1083},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 14), Port: duplicatedPort},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 15), Port: duplicatedPort},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			iptablesScanner := fakeScanner{
+				expectedEntries: tt.expectedEntries,
+				expectedErr:     tt.expectedAddFuncErr,
+			}
+
+			testTracker := fakeTracker{
+				receivedID:          make(chan string),
+				receivedRemoveID:    make(chan string),
+				receivedPortMapping: make(chan nat.PortMap),
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			interval := time.Second
+			iptablesHandler := iptables.New(ctx, &testTracker, &iptablesScanner, tt.listenerIP, interval)
+
+			go func() {
+				require.NoError(t, iptablesHandler.ForwardPorts())
+				cancel()
+			}()
+
+			for i := 0; i < len(tt.expectedEntries); i++ {
+				id := <-testTracker.receivedID
+				expectedID := utils.GenerateID(entryToString(tt.expectedEntries[i]))
+				require.Equal(t, expectedID, id)
+
+				pm := <-testTracker.receivedPortMapping
+				portProto, err := nat.NewPort("tcp", strconv.Itoa(tt.expectedEntries[i].Port))
+				require.NoError(t, err)
+
+				// Port bindings for the same port on different IP addresses should appear only once
+				// in the port mapping. This is because the HostIP is always controlled by the
+				// k8sServiceListenerAddr, which means that duplicate entries with the same port
+				// but different IPs are unnecessary and should not be handled.
```

Comment:
This is only true for the iptables scanner though. It is possible that the same port may be forwarded from both the proc/net and the iptables scanners, and the first one "wins".

I wonder if the right thing happens when the first instance of the duplicate port is closed; will the forwarding remain until the original port on the other IP address is closed?

Reply:
Just to clarify, the system operates exclusively on either `127.0.0.1` or `0.0.0.0`. The distinction here is that we have a configuration option (k8sServiceListenerAddr) for specifying the address, while in other places, the address is provided by user input (e.g., `docker run -it -p 127.0.0.1:8080:80`).
Regarding port duplication, the behavior varies depending on the underlying API. For example:
Docker: If a user attempts to bind a port (e.g., port 80) already in use (e.g., by iptables), Docker will return an error via the CLI.
Containerd: In contrast, Containerd won’t return an error; instead, it will silently fail without any notification (user has to look through the guestAgent logs to see the failure)
Docker/Containerd with --net=host: When using the `--net=host` option, Docker/containerd rely on the /proc/net scanner. If the port is already bound to an IP, Docker will display an error, but in the case of Containerd, the failure will be silent, with no error returned(again there will be a log entry in guestAgent).
In Kubernetes, if a service tries to bind a port that’s already in use, an error will be generated during service creation if the port is bound via the API. However, iptables is only involved with Kubernetes services that don't go through the API. For instance, if iptables detects that port 80 is already occupied (say, by a Docker proxy), Kubernetes will fail to bind the port to an IP. This will result in an error log in the guestAgent logs, but the binding won't be successful.